core/src/main/scala/cats/TraverseFilter.scala

```diff
@@ -140,13 +140,23 @@ trait TraverseFilter[F[_]] extends FunctorFilter[F] {
    * Removes duplicate elements from a list, keeping only the first occurrence.
    * This is usually faster than ordDistinct, especially for things that have a slow comparion (like String).
    */
-  def hashDistinct[A](fa: F[A])(implicit H: Hash[A]): F[A] =
-    traverseFilter[State[HashSet[A], *], A, A](fa)(a =>
-      State(alreadyIn => if (alreadyIn(a)) (alreadyIn, None) else (alreadyIn + a, Some(a)))
-    )
+  def hashDistinct[A](fa: F[A])(implicit H: Hash[A]): F[A] = {
+
+    final class Wrapper(val value: A) {
```

Comment:
idea: what if we did:
```scala
trait Hash[A] extends Eq[A] {
  class Wrapper private (val unwrap: A) {
    override def equals(x: Any): Boolean =
      x match {
        case that: Wrapper =>
          eqv(unwrap, that.unwrap)
        case _ => false
      }

     override val hashCode: Int = hash(unwrap) 
  }

  def wrap(a: A): Wrapper = new Wrapper(a)
}
```

Then we could use `H.wrap(a)` below and this same pattern could be useful for others using JVM interop.

I don't know what the binary compatibility story is on adding member classes like this though.

Reply:
I checked and it's binary-compatible. But it feels a bit ad-hoc. Should we add similar wrappers for `Eq` and `Order`?

Reply:
It may be ad-hoc but there are two important differences:

1. scala doesn't use typeclasses for hashing but did that with Ordering.
2. hashCode and equals are so deeply built into the JVM that many use cases assume that ability.

Eq is somewhat similar, but less impactful IMO due to the outsized role of hashing.

Reply:
Another question: isn't it awkward to work with a type defined like this? Since it's not a static type, but a member of a specific instance of the typeclass.

Reply:
It's a path dependent type, which a lot of scala users almost pretend don't exist but they are quite powerful.

I've used them in several cases an can give some really nice invariants. In this case it is exactly what we want: statically guaranteeing that two values wrap something using the same Hash value.

But anyway, it's just a suggestion. We could add it later. You could make a more minimal change by just using your code (which is basically inlining this design at each use site, but there is only one use site so, it's fine)

Reply:
Just to clarify, since we're wrapping each `a` into `Wrapper`, it should have some impact on the overall performance. But we're ok with it for now – is it correct?